### DISCUSSION TITLE
Dimmable LightBulb with ColorTemperature selection

### DISCUSSION THREAD

**Discussion by chellmann** [HomeSpan/HomeSpan#82] [Q&A]
Hello,

how could one configure a LightBulb with Warm and Cool-White PWM Channels? I have read through the Examples and Docs but only found the Hue/Saturation options for RGB.

Thanks a lot!

**Comment by HomeSpan** [comment]
To implement a color-temperature light bulb you would use the following:

```C++
new Service::LightBulb();
  new Characteristic::On(); 
  new Characteristic::Brightness();
  new Characteristic::ColorTemperature();
```

Note that Characteristic::Saturation() and Characteristic::Hue() cannot be used at the same time as Characteristic::ColorTemperature().  The first two are used for RGB LEDs whereas Characteristic::ColorTemperature() is used for setting the "temperature" of a bulb (e.g. 2000K, 4000K, etc.).

**Reply by chellmann** [reply]
Thanks a lot. I tried to implement this by altering the RGB-LED sample.
In the Home App i can not Pair the device. The App tells "Device is out of compliance",
in Debug Level 2, the Board tells me *** ACCESSORY PAIRED! ***

The send JSON String is:
`{"accessories":[{"aid":1,"services":[{"iid":1,"type":"3E","characteristics":[{"iid":2,"type":"23","value":"Bridge #1","format":"string","perms":["pr"]},{"iid":3,"type":"20","value":"HomeSpan","format":"string","perms":["pr"]},{"iid":4,"type":"30","value":"123-ABC","format":"string","perms":["pr"]},{"iid":5,"type":"21","value":"HS Bridge","format":"string","perms":["pr"]},{"iid":6,"type":"52","value":"0.9","format":"string","perms":["pr"]},{"iid":7,"type":"14","format":"bool","perms":["pw"]}]},{"iid":8,"type":"A2","characteristics":[{"iid":9,"type":"37","value":"1.1.0","format":"string","perms":["pr"]}]}]},{"aid":2,"services":[{"iid":1,"type":"3E","characteristics":[{"iid":2,"type":"23","value":"On/Off LED","format":"string","perms":["pr"]},{"iid":3,"type":"20","value":"HomeSpan","format":"string","perms":["pr"]},{"iid":4,"type":"30","value":"123-ABC","format":"string","perms":["pr"]},{"iid":5,"type":"21","value":"20mA LED","format":"string","perms":["pr"]},{"iid":6,"type":"52","value":"0.9","format":"string","perms":["pr"]},{"iid":7,"type":"14","format":"bool","perms":["pw"]}]},{"iid":8,"type":"43","characteristics":[{"iid":9,"type":"25","value":0,"format":"bool","perms":["pr","pw","ev"]}]}]},{"aid":3,"services":[{"iid":1,"type":"3E","characteristics":[{"iid":2,"type":"23","value":"CW WW LED","format":"string","perms":["pr"]},{"iid":3,"type":"20","value":"HomeSpan","format":"string","perms":["pr"]},{"iid":4,"type":"30","value":"123-ABC","format":"string","perms":["pr"]},{"iid":5,"type":"21","value":"20mA LED","format":"string","perms":["pr"]},{"iid":6,"type":"52","value":"0.9","format":"string","perms":["pr"]},{"iid":7,"type":"14","format":"bool","perms":["pw"]}]},{"iid":8,"type":"43","characteristics":[{"iid":9,"type":"25","value":0,"format":"bool","perms":["pr","pw","ev"]},{"iid":10,"type":"CE","value":0,"format":"uint32","perms":["pr","pw","ev"]},{"iid":11,"type":"8","value":100,"format":"int","minValue":5,"maxValue":100,"minStep":1,"perms":["pr","pw","ev"]}]}]}]}`

By removing the Device with ColorTemperature() i managed to pair the device successfully. Any clues?

**Comment by HomeSpan** [comment]
Yes, based on above it appears you have set the ColorTemperature to 0 ("type:"CE","value":0).  This is an impermissible value and will cause the Home App error you have received.  The valid range for ColorTemperature is 140-500.  If you instantiate ColorTemperature without any arguments it will default to 200.

HomeSpan should have thrown a warning upon startup indicating this error.  Something like this:

```
*** HomeSpan Info ***

➤ Accessory:  AID=1
   ➟ Service AccessoryInformation:  IID=1, UUID=0x3E
      ⇨ Characteristic Name("HomeSpan Test"):  IID=2, UUID=0x23
      ⇨ Characteristic Manufacturer("HomeSpan"):  IID=3, UUID=0x20
      ⇨ Characteristic SerialNumber("HSL-123"):  IID=4, UUID=0x30
      ⇨ Characteristic Model("HSL Test"):  IID=5, UUID=0x21
      ⇨ Characteristic FirmwareRevision("1.2.1"):  IID=6, UUID=0x52
      ⇨ Characteristic Identify(0):  IID=7, UUID=0x14
   ➟ Service HAPProtocolInformation:  IID=8, UUID=0xA2
      ⇨ Characteristic Version("1.1.0"):  IID=9, UUID=0x37
   ➟ Service LightBulb:  IID=10, UUID=0x43
      ⇨ Characteristic On(0):  IID=11, UUID=0x25
      ⇨ Characteristic Brightness(0):  IID=12, UUID=0x8  Range=[0,100]
      ⇨ Characteristic Name("Light 1"):  IID=13, UUID=0x23
      ⇨ Characteristic ColorTemperature(0):  IID=14, UUID=0xCE  Range=[140,500]
   ➟ Service LightBulb:  IID=15, UUID=0x43
      ⇨ Characteristic On(1):  IID=16, UUID=0x25
      ⇨ Characteristic Brightness(50):  IID=17, UUID=0x8  Range=[0,100]
         ⬌ Set Range for Brightness with IID=17: Min=10, Max=100, Step=5
      ⇨ Characteristic Name("Light 2"):  IID=18, UUID=0x23
      ✘ Characteristic ColorTemperature with IID=14  *** WARNING: Initial value of 0 is out of range [140,500]. ***

*** CAUTION: There is 1 WARNING associated with this configuration that may lead to the device becoming non-responsive, or operating in an unexpected manner. ***
```

Did this warning appear?

Note the HAP-R2 documentation seems incorrect and shows valid ranges of 50-400.  HomeSpan uses the correct range 140-500.  Also note that for some reason Apple decided to measure color temperature in "inverse" kelvins, defined as 1,000,000 / temp.  The range 140-500 maps to 7,142K through 2,000K, which should cover all light bulbs (typical range is 2,000K-5000K).

**Reply by chellmann** [reply]
Yes, that warning did appear, but i didn't read it.
Now with this changes the project works like a charm. I really like this project and want to thanks all collaborators.
